bser.c

```diff
@@ -0,0 +1,366 @@
+}
+
+const char* bser_error_message(bser_t* bser)
+{
+    assert(bser_is_error(bser));
+    return bser->value.error_message;
+}
+
+bser_t*
+bser_new_integer(uint64_t value, bser_t* fill)
+{
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_INT64;
+    fill->value.integer = value;
+    return fill;
+}
+
+bser_t*
+bser_new_real(double value, bser_t* fill)
+{
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_REAL;
+    fill->value.real = value;
+    return fill;
+}
+
+bser_t*
+bser_new_true(bser_t* fill)
+{
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_TRUE;
+    return fill;
+}
+
+bser_t*
+bser_new_false(bser_t* fill)
+{
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_FALSE;
+    return fill;
+}
+
+bser_t*
+bser_new_null(bser_t* fill)
+{
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_NULL;
+    return fill;
+}
+
+bser_t*
+bser_new_string(const char* chars, size_t len, bser_t* fill)
+{
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_STRING;
+    fill->value.string.chars = chars;
+    fill->value.string.length = len;
+    return fill;
+}
+
+bser_t*
+bser_new_array(bser_t* elems, size_t len, bser_t* fill)
+{
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_ARRAY;
+    fill->value.array.elements = elems;
+    fill->value.array.length = len;
+    return fill;
+}
+
+bser_t*
+bser_new_object(bser_key_value_pair_t* fields,
+        size_t length, bser_t* fill)
+{
+    if (fill == NULL) {
+        fill = bser_alloc();
+    }
+    fill->type = BSER_TAG_OBJECT;
+    fill->value.object.fields = fields;
+    fill->value.object.length = length;
+    return fill;
+}
+
+static void
+bser_free_array_contents(bser_t* bser)
+{
+    assert(bser_is_array(bser));
+    for (int i = 0; i < bser->value.array.length; ++i) {
+        bser_free_contents(&bser->value.array.elements[i]);
+    }
+    free(bser->value.array.elements);
+}
+
+static void
+bser_free_object_contents(bser_t* bser)
+{
+    assert(bser_is_object(bser));
+    for (int i = 0; i < bser->value.object.length; ++i) {
+        bser_free_contents(&bser->value.object.fields[i].value);
+    }
+    free(bser->value.object.fields);
+}
+
+void
+bser_free_contents(bser_t* bser)
+{
+    if (bser_is_array(bser)) {
+        bser_free_array_contents(bser);
+    } else if (bser_is_object(bser)) {
+        bser_free_object_contents(bser);
+    }
+    bser->type = BSER_TAG_ERROR;
+    bser->value.error_message = "<deallocated>";
+}
+
+void
+bser_free(bser_t* bser) {
+    bser_free_contents(bser);
+    free(bser);
+}
+
+int
+bser_string_strcmp(const char* match, bser_t* bser)
+{
+    assert(bser_is_string(bser));
+    bser_parse_if_necessary(bser);
+    int cmp = memcmp(match, bser->value.string.chars, bser->value.string.length);
```

Comment:
Well, memcmp will return the right value, as there will be a \0 in position match[strlen(match)].  However, memcmp() may read beyond the end of the string.  I'll address this.